scripts/build_go_matrix.sh

```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env bash
+
+# This script builds multiple Go binaries for Coder with the given OS and
+# architecture combinations.
+#
+# Usage: ./build_go_matrix.sh [--version 1.2.3+devel.abcdef] [--output dist/] [--slim] [--sign-darwin] [--archive] [--package-linux] os1:arch1,arch2 os2:arch1 os1:arch3
+#
+# If no OS:arch combinations are provided, nothing will happen and no error will
+# be returned. Slim builds are disabled by default. If no version is specified,
+# defaults to the version from ./version.sh
+#
+# The --output parameter must be a directory with a trailing slash where all
+# files will be dropped with the default name scheme
+# `coder_$version_$os_$arch(.exe)?`, or must contain the `{os}` and `{arch}`
+# template variables. You may also use `{version}`. Note that for windows builds
+# the `.exe` suffix will be appended automatically.
+#
+# Unless overridden via --output, the built binary will be dropped in
+# "$repo_root/dist/coder_$version_$os_$arch" (with a ".exe" suffix for windows
+# builds).
+#
+# If the --sign-darwin parameter is specified, all darwin binaries will be
+# signed using the `codesign` utility. $AC_APPLICATION_IDENTITY must be set and
+# the signing certificate must be imported for this to work.
+#
+# If the --archive parameter is specified, all binaries will be archived using
+# ./archive.sh. The --sign-darwin parameter will be carried through, and all
+# archive files will be dropped in the output directory with the same name as
+# the binary and the .zip (for windows and darwin) or .tar.gz extension.
+#
+# If the --package-linux parameter is specified, all linux binaries will be
+# packaged using ./package.sh. Requires the nfpm binary.
+
+set -euo pipefail
+# shellcheck source=scripts/lib.sh
+source "$(dirname "${BASH_SOURCE[0]}")/lib.sh"
+
+version=""
+output_path=""
+slim=0
+sign_darwin=0
+archive=0
+package_linux=0
+
+args="$(getopt -o "" -l version:,output:,slim,sign-darwin,archive,package-linux -- "$@")"
+eval set -- "$args"
+while true; do
+	case "$1" in
+	--version)
+		version="$2"
+		shift 2
+		;;
+	--output)
+		output_path="$2"
+		shift 2
+		;;
+	--slim)
+		slim=1
+		shift
+		;;
+	--sign-darwin)
+		if [[ "${AC_APPLICATION_IDENTITY:-}" == "" ]]; then
+			error "AC_APPLICATION_IDENTITY must be set when --sign-darwin is supplied"
+		fi
+		sign_darwin=1
+		shift
+		;;
+	--archive)
+		archive=1
+		shift
+		;;
+	--package-linux)
+		package_linux=1
+		shift
+		;;
+	--)
+		shift
+		break
+		;;
+	*)
+		error "Unrecognized option: $1"
+		;;
+	esac
+done
+
+# Verify the output path template.
+if [[ "$output_path" == "" ]]; then
+	# Input paths are relative, so we don't cdroot at the top, but for this case
+	# we want it to be relative to the root.
+	cdroot
+	mkdir -p dist
+	output_path="$(realpath "dist/coder_{version}_{os}_{arch}")"
```

Comment:
Won't `realpath` error out here if the coder folder doesn't exist (since we're using `set -e`)? Maybe it's supposed to exist but then the `mkdir` above seems redundant. (Similar cases in the slim build script.)

Reply:
The binaries aren't output into directories anymore, and are instead just put directly into the `dist` dir. Whatever path you give to realpath has to have a `dirname` that exists otherwise it fails